flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/sink/dynamic/DynamicRecordProcessor.java

```diff
@@ -0,0 +1,163 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.sink.dynamic;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.functions.OpenContext;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.api.java.tuple.Tuple3;
+import org.apache.flink.streaming.api.functions.ProcessFunction;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.util.Collector;
+import org.apache.flink.util.OutputTag;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.catalog.Catalog;
+import org.apache.iceberg.flink.CatalogLoader;
+
+@Internal
+class DynamicRecordProcessor<T> extends ProcessFunction<T, DynamicRecordInternal>
+    implements Collector<DynamicRecord> {
+  static final String DYNAMIC_TABLE_UPDATE_STREAM = "dynamic-table-update-stream";
+
+  private final DynamicRecordGenerator<T> generator;
+  private final CatalogLoader catalogLoader;
+  private final boolean immediateUpdate;
+  private final int cacheMaximumSize;
+  private final long cacheRefreshMs;
+
+  private transient TableMetadataCache tableCache;
+  private transient HashKeyGenerator hashKeyGenerator;
+  private transient TableUpdater updater;
+  private transient OutputTag<DynamicRecordInternal> updateStream;
+  private transient Collector<DynamicRecordInternal> collector;
+  private transient Context context;
+
+  DynamicRecordProcessor(
+      DynamicRecordGenerator<T> generator,
+      CatalogLoader catalogLoader,
+      boolean immediateUpdate,
+      int cacheMaximumSize,
+      long cacheRefreshMs) {
+    this.generator = generator;
+    this.catalogLoader = catalogLoader;
+    this.immediateUpdate = immediateUpdate;
+    this.cacheMaximumSize = cacheMaximumSize;
+    this.cacheRefreshMs = cacheRefreshMs;
+  }
+
+  @Override
+  public void open(OpenContext openContext) throws Exception {
+    super.open(openContext);
+    Catalog catalog = catalogLoader.loadCatalog();
+    this.tableCache = new TableMetadataCache(catalog, cacheMaximumSize, cacheRefreshMs);
+    this.hashKeyGenerator =
+        new HashKeyGenerator(
+            cacheMaximumSize, getRuntimeContext().getTaskInfo().getMaxNumberOfParallelSubtasks());
+    if (immediateUpdate) {
+      updater = new TableUpdater(tableCache, catalog);
+    }
+
+    updateStream =
```

Comment:
is this in `else`?

Reply:
I mean `if (immediateUpdate) {} else {...}`

Reply:
No, this is correct. TableUpdater won't be used if we go through the DynamicTableUpdateOperator (immediateUpdate == false).

Reply:
Yeah, but why do we create the object, and the cache, and everything?

Reply:
Do we need the `updateStream` if the `immediateUpdate` is `false`

Reply:
Oh yes, you are right. See https://github.com/apache/iceberg/pull/13304/commits/8284ef17c28f431e11639f42d72242a27af46781